Milliner/src/Service/MillinerService.php

```diff
@@ -20,230 +20,808 @@ class MillinerService implements MillinerServiceInterface
+     *
+     * @param $jsonld
+     * @param $predicate
+     * @param $value
+     *
+     * @return mixed string|null
+     */
+    protected function getFirstPredicate(array $jsonld, $predicate, $value = true)
+    {
+        $key = $value ? '@value' : '@id';
+        $malformed = empty($jsonld) ||
+            !isset($jsonld[0][$predicate]) ||
+            empty($jsonld[0][$predicate]) ||
+            !isset($jsonld[0][$predicate][0][$key]);
+
+        if ($malformed) {
+            return null;
+        }
+
+        return $jsonld[0][$predicate][0][$key];
+    }
+
+    /**
+     * Extracts a modified date from jsonld and returns it as a timestamp.
+     *
+     * @param array $jsonld
+     *
+     * @return int
+     *
+     * @throws \RuntimeException
+     */
+    protected function getModifiedTimestamp(array $jsonld)
+    {
+        $modified = $this->getFirstPredicate(
+            $jsonld,
+            $this->modifiedDatePredicate
+        );
+
+        if (empty($modified)) {
+            throw new \RuntimeException(
+                "Could not parse {$this->modifiedDatePredicate} from " . json_encode($jsonld),
+                500
+            );
+        }
+
+        $date = \DateTime::createFromFormat(
+            \DateTime::W3C,
+            $modified
+        );
+
+        return $date->getTimestamp();
     }
 
     /**
      * {@inheritDoc}
      */
-    public function updateRdf(
-        $drupal_jsonld,
-        $drupal_path,
-        $token
+    public function saveMedia(
+        $json_url,
+        $jsonld_url,
+        $token = null
     ) {
-        $fedora_path = $this->pathMapper->getFedoraPath($drupal_path);
-        if ($fedora_path === null) {
+        // First get the media json from Drupal.
+        $headers = empty($token) ? [] : ['Authorization' => $token];
+        $drupal_response = $this->drupal->get(
+            $json_url,
+            ['headers' => $headers]
+        );
+
+        $media_json = json_decode(
+            $drupal_response->getBody(),
+            true
+        );
+
+        // Extract the file uuid.  It can be under 'field_file' or 'field_image'.
+        if (!isset($media_json['field_image']) && !isset($media_json['field_file'])) {
+            throw new \RuntimeException(
+                "Cannot parse file UUID from $json_url.  Media must use 'field_file' or 'field_image'.",
+                500
+            );
+        }
+
+        $field_name = isset($media_json['field_image']) ? 'field_image' : 'field_file';
+
+        if (empty($media_json[$field_name])) {
+            throw new \RuntimeException(
+                "Cannot parse file UUID from $json_url.  'field_file' or 'field_image' is empty.",
+                500
+            );
+        }
+
+        $file_uuid = $media_json[$field_name][0]['target_uuid'];
+
+        // Get the file's LDP-NR counterpart in Fedora.
+        $urls = $this->gemini->getUrls($file_uuid, $token);
+
+        if (empty($urls)) {
```

Comment:
I'm actually never mapping Media in Gemini, just Files and Content.  I'm looking up the Media in a roundabout way since timing issues can leave me in a scenario where I can't link everything together, particuarly when deleting.  I was winding up with dangling references to Media that no longer existed in some corner cases.  It's pretty extreme, but when doing strange things like creating a File and Media, deleting the Media, and then creating a new Media and associating it with the original file you can run into issues like this.

So I'm just always looking it up every time to guarantee I get it right.  It's slower but guaranteed to work.

FYI, https://github.com/Islandora-CLAW/CLAW/issues/721 would cut down on some of the hassle when doing the lookup and provide some caching.